### PR TITLE
feat: add lattice_reduce and lattice_fold

### PR DESCRIPTION
**Pull request by zzlk**
fixes https://github.com/hydro-project/hydroflow/issues/752

### CODE REVIEW

File: hydroflow_lang/src/graph/ops/lattice_reduce.rs [hunk 1]
```diff
@@ -14,22 +14,26 @@ use crate::graph::ops::OperatorWriteOutput;
...
+/// `lattice_reduce` can also be provided with one generic lifetime persistence argument, either
 /// `'tick` or `'static`, to specify how data persists. With `'tick`, values will only be collected
 /// within the same tick. With `'static`, values will be remembered across ticks and will be
 /// aggregated with pairs arriving in later ticks. When not explicitly specified persistence
 /// defaults to `'static`.
 ///
+/// `lattice_reduce` is differentiated from `lattice_fold` in that `lattice_reduce` the accumulating type does not need to have a sensible default.
+/// But it also means that the accumulating function inputs and the accumulating type must be the same.
+///
+/// TODO: give example here that shows you don't need to implement Default
```
Comment (zzlk) [inline hydroflow_lang/src/graph/ops/lattice_reduce.rs:28]: missed this one.

File: hydroflow_lang/src/graph/ops/lattice_reduce.rs [hunk 2]
```diff
@@ -14,22 +14,25 @@ use crate::graph::ops::OperatorWriteOutput;
...
+/// but specialized for lattice types. `lattice_reduce::<MyLattice>()` is equivalent to `reduce(hydroflow::lattices::Merge::merge_owned)`.
 ///
-/// `lattice_merge` can also be provided with one generic lifetime persistence argument, either
+/// `lattice_reduce` can also be provided with one generic lifetime persistence argument, either
 /// `'tick` or `'static`, to specify how data persists. With `'tick`, values will only be collected
 /// within the same tick. With `'static`, values will be remembered across ticks and will be
 /// aggregated with pairs arriving in later ticks. When not explicitly specified persistence
 /// defaults to `'static`.
 ///
+/// `lattice_reduce` is differentiated from `lattice_fold` in that `lattice_reduce` the accumulating type does not need to have a sensible default.
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/ops/lattice_reduce.rs:25]: "sensible default" = implements `Default`?

File: hydroflow_lang/src/graph/ops/lattice_fold.rs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+/// but specialized for lattice types. `lattice_fold::<MyLattice>()` is equivalent to `fold(MyLattice::default(), hydroflow::lattices::Merge::merge_owned)`.
+///
+/// `lattice_fold` can also be provided with one generic lifetime persistence argument, either
+/// `'tick` or `'static`, to specify how data persists. With `'tick`, values will only be collected
+/// within the same tick. With `'static`, values will be remembered across ticks and will be
+/// aggregated with pairs arriving in later ticks. When not explicitly specified persistence
+/// defaults to `'static`.
+///
+/// `lattice_fold` is differentiated from `lattice_reduce` in that `lattice_fold` can accumulate into a different type from its input.
+/// But it also means that the accumulating type must have a sensible default value.
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/ops/lattice_fold.rs:24]: "must implement `Default`"?

File: hydroflow_lang/src/graph/ops/lattice_fold.rs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                       },
+                   ..
+               },
+               diagnostics| {
+        assert!(is_pull);
+
+        let lat_type = &type_args[0];
+
+        let arguments = parse_quote_spanned! {lat_type.span()=> // Uses `lat_type.span()`!
+            <#lat_type>::default(), #root::lattices::Merge::merge_owned
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/ops/lattice_fold.rs:66]: `<#lat_type as ::std::default::Default>::default()`

**Review by MingweiSamuel — APPROVED** [review]
lgtm

File: hydroflow_lang/src/graph/ops/lattice_fold.rs (hunk 4 above)
Comment (zzlk) [reply hydroflow_lang/src/graph/ops/lattice_fold.rs:66]: hmm, I wonder what is the difference in this case?

**Pull request merged**